File: PassJs.Web/PassJs.Web/ClientApp/src/app/models/sub-task.ts
import { Mapper } from "../utils/helpers";
import { AnswerSubTaskOptionViewModel } from "./answersubtaskoption";

export enum SubTaskType {
    Code = 0,
    Test,
    LongAnswer
}

export enum SubTaskLevel {
    Junior = 0,
    Middle,
    Senior
}

export class SubTaskViewModel {
    constructor(model?: SubTaskViewModel) {
        if (model) {
            Mapper.map(model, this);

            if (this.type === SubTaskType.Test && model.answerSubTaskOptions) {
                this.answerSubTaskOptions = model.answerSubTaskOptions.map(a => new AnswerSubTaskOptionViewModel(a));
            }
        }
    }

    id: number;
    taskHeadId: number;
    text: string;
    taskText: string;
    answer: string;
    unitTestsCode: string;
    reporterCode: string;
    title: string;
    order: number;
    published: boolean;
    publishNow: boolean;
    type: SubTaskType;
    level: SubTaskLevel;
    answerSubTaskOptions: AnswerSubTaskOptionViewModel[] = [];
}

File: CodeSchool.DomainModels/SubTask.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace PassJs.DomainModels
{
    public enum SubTaskType
    {
        Code = 0,
        Test,
        LongAnswer
    }

    public enum SubTaskLevel
    {
        Junior = 0,
        Middle,
        Senior
    }

    public class SubTask
    {
        public SubTask()
        {
            AnswerSubTaskOptions = new List<AnswerSubTaskOption>();
            UserSubTasks = new List<UserSubTask>();
        }

        [Required]
        public int Id { get; set; }

        [Required]
        public int TaskHeadId { get; set; }

        [Required]
        public string Title { get; set; }

        [Required]
        public Guid CompanyId { get; set; }

        public string Text { get; set; }

        public string Answer { get; set; }

        public string TaskText { get; set; }

        public string UnitTestsCode { get; set; }

        public string ReporterCode { get; set; }

        public SubTaskType Type { get; set; }

        public SubTaskLevel Level { get; set; }

        public int Order { get; set; }

        public bool Published { get; set; }

        public virtual TaskHead TaskHead { get; set; }

        public virtual ICollection<UserSubTask> UserSubTasks { get; set; }

        public virtual ICollection<AnswerSubTaskOption> AnswerSubTaskOptions { get; set; }
    }
}
